[Maint] The Gallery is broken in CircleCI previews

## 🧰 Task

Figure out what the issue is, the CircleCI action is different than the build_docs action, which works correctly—the gallery is fine in the .zip.
There is a difference in the order of the installs, which results in different ninja version.
There is a difference in python version
There is a difference in Qt backend.

I took some stabs at the first and second part, but that resulted in worse outcomes, with one of the examples failing totally and the build failing as as result, see:
https://github.com/napari/docs/pull/173

Ok! Currently, changing the system/qt dependencies for CircleCI the error has moved from the one above to 
```
generating gallery for gallery... [ 91%] nD_multiscale_image_non_uniform.py
Killed
make: *** [Makefile:30: docs-build] Error 137

Exited with code exit status 2
CircleCI received exit code 2
```

As @psobolewskiPhD [has found](https://github.com/napari/docs/pull/178#issuecomment-1594232748), this may mean CircleCI is running out of memory: https://support.circleci.com/hc/en-us/articles/115014359648-Exit-Code-137-Out-of-Memory

I tried increasing the CircleCI image memory but we don't have that machine available in our tier. So currently our options are:
- Increase the CircleCI memory tier (by paying)
- Figuring out why the gallery is consuming so much memory.
I am hoping maybe one of the Qt experts may help with finding a way to either clean up after each example or make sure we handle this better? 

Feel free to inspect things here: https://github.com/napari/docs/pull/178

--- a/docs/conf.py
+++ b/docs/conf.py
@@ -19,8 +19,9 @@ from importlib import import_module
 from pathlib import Path
 from urllib.parse import urlparse, urlunparse
 
-import qtgallery
 from jinja2.filters import FILTERS
+from sphinx_gallery import scrapers
+from sphinx_gallery.sorting import ExampleTitleSortKey
 
 import napari
 from napari._version import __version_tuple__
@@ -164,12 +165,12 @@ python_version = '3.9'
 python_version_range = '3.8–3.10'
 
 myst_substitutions = {
-   "napari_conda_version": f"`napari={version_string}`",
-   "napari_version": version_string,
-   "python_version": python_version,
-   "python_version_range": python_version_range,
-   "python_version_code": f"`python={python_version}`",
-   "conda_create_env": f"```sh\nconda create -y -n napari-env -c conda-forge python={python_version}\nconda activate napari-env\n```",
+    "napari_conda_version": f"`napari={version_string}`",
+    "napari_version": version_string,
+    "python_version": python_version,
+    "python_version_range": python_version_range,
+    "python_version_code": f"`python={python_version}`",
+    "conda_create_env": f"```sh\nconda create -y -n napari-env -c conda-forge python={python_version}\nconda activate napari-env\n```",
 }
 
 nb_output_stderr = 'show'
@@ -197,19 +198,50 @@ exclude_patterns = [
 napoleon_custom_sections = [('Events', 'params_style')]
 
 
-def reset_napari_theme(gallery_conf, fname):
+def reset_napari(gallery_conf, fname):
     from napari.settings import get_settings
+    from qtpy.QtWidgets import QApplication
 
     settings = get_settings()
     settings.appearance.theme = 'dark'
-    qtgallery.reset_qapp(gallery_conf, fname)
+
+    # Disabling `QApplication.exec_` means example scripts can call `exec_`
+    # (scripts work when run normally) without blocking example execution by
+    # sphinx-gallery. (from qtgallery)
+    QApplication.exec_ = lambda _: None
 
 
-from sphinx_gallery.sorting import ExampleTitleSortKey
+def napari_scraper(block, block_vars, gallery_conf):
+    """Basic napari window scraper.
+
+    Looks for any QtMainWindow instances and takes a screenshot of them.
+
+    `app.processEvents()` allows Qt events to propagateo and prevents hanging.
+    """
+    imgpath_iter = block_vars['image_path_iterator']
+
+    if app := napari.qt.get_app():
+        app.processEvents()
+    else:
+        return ""
+
+    img_paths = []
+    for win, img_path in zip(
+        reversed(napari._qt.qt_main_window._QtMainWindow._instances),
+        imgpath_iter,
+    ):
+        img_paths.append(img_path)
+        win._window.screenshot(img_path, canvas_only=False)
+
+    napari.Viewer.close_all()
+    app.processEvents()
+
+    return scrapers.figure_rst(img_paths, gallery_conf['src_dir'])
+
 
 sphinx_gallery_conf = {
-    #'examples_dirs': '../../napari/examples',  # path to your example scripts
-                                                # this value is set in the Makefile
+    # path to your example scripts (this value is set in the Makefile)
+    # 'examples_dirs': '../../napari/examples',
     'gallery_dirs': 'gallery',  # path to where to save gallery generated output
     'filename_pattern': '/*.py',
     'ignore_pattern': 'README.rst|/*_.py',
@@ -218,8 +250,8 @@ sphinx_gallery_conf = {
     'download_all_examples': False,
     'min_reported_time': 10,
     'only_warn_on_example_error': True,
-    'image_scrapers': ("matplotlib", qtgallery.qtscraper,),
-    'reset_modules': (reset_napari_theme,),
+    'image_scrapers': ("matplotlib", napari_scraper,),
+    'reset_modules': (reset_napari,),
     'reference_url': {'napari': None},
     'within_subsection_order': ExampleTitleSortKey,
 }
@@ -262,7 +294,7 @@ linkcheck_ignore = [
     'https://napari.zulipchat.com/',
     '../_tags',
     'https://en.wikipedia.org/wiki/Napari#/media/File:Tabuaeran_Kiribati.jpg',
-    ]
+]
 
 
 def rewrite_github_anchor(app, uri: str):

--- a/docs/requirements.txt
+++ b/docs/requirements.txt
@@ -10,6 +10,5 @@ sphinx_autodoc_typehints==1.12.0
 myst-nb
 napari-sphinx-theme
 matplotlib
-qtgallery
 lxml
 imageio-ffmpeg

--- a/requirements.txt
+++ b/requirements.txt
@@ -11,6 +11,5 @@ sphinx_autodoc_typehints==1.12.0
 myst-nb
 napari-sphinx-theme
 matplotlib
-qtgallery
 lxml
 imageio-ffmpeg

